File: python/connect/connect.py
OFFSETS = [(-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0)]


class ConnectGame:
    def __init__(self, board):
        self.board = [line.replace(' ', '') for line in board.splitlines()]

    def get_winner(self):
        row = len(self.board)
        col = len(self.board[0])

        for c in range(col):
            if self.board[0][c] == 'O' and any(map(lambda cell: cell[0] == row - 1, self.find_connected_cells(0, c))):
                return 'O'

        for r in range(row):
            if self.board[r][0] == 'X' and any(map(lambda cell: cell[1] == col - 1, self.find_connected_cells(r, 0))):
                return 'X'

        return ''

    def find_connected_cells(self, r, c):
        connected_cells = set()
        self.search(connected_cells, r, c, self.board[r][c])
        return connected_cells

    def search(self, connected_cells, r, c, target):
        row = len(self.board)
        col = len(self.board[0])

        connected_cells.add((r, c))

        for r_offset, c_offset in OFFSETS:
            next_r, next_c = r + r_offset, c + c_offset
            if 0 <= next_r < row and 0 <= next_c < col and self.board[next_r][next_c] == target and (next_r, next_c) not in connected_cells:
                self.search(connected_cells, next_r, next_c, target)

File: java/connect/src/main/java/Connect.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Connect {
  static final int[] R_OFFSETS = {-1, -1, 0, 0, 1, 1};
  static final int[] C_OFFSETS = {0, 1, -1, 1, -1, 0};

  String[] board;

  public Connect(String[] board) {
    this.board = Arrays.stream(board).map(line -> line.replace(" ", "")).toArray(String[]::new);
  }

  public Winner computeWinner() {
    int row = board.length;
    int col = board[0].length();

    for (int c = 0; c < col; ++c) {
      if (board[0].charAt(c) == 'O'
          && findConnectedCells(0, c).stream().anyMatch(p -> p.r() == row - 1)) {
        return Winner.PLAYER_O;
      }
    }

    for (int r = 0; r < row; ++r) {
      if (board[r].charAt(0) == 'X'
          && findConnectedCells(r, 0).stream().anyMatch(p -> p.c() == col - 1)) {
        return Winner.PLAYER_X;
      }
    }

    return Winner.NONE;
  }

  Set<Point> findConnectedCells(int r, int c) {
    Set<Point> connectedCells = new HashSet<>();
    search(connectedCells, r, c);

    return connectedCells;
  }

  void search(Set<Point> connectedCells, int r, int c) {
    int row = board.length;
    int col = board[0].length();

    connectedCells.add(new Point(r, c));

    for (int i = 0; i < R_OFFSETS.length; ++i) {
      int adjR = r + R_OFFSETS[i];
      int adjC = c + C_OFFSETS[i];
      if (adjR >= 0
          && adjR < row
          && adjC >= 0
          && adjC < col
          && board[adjR].charAt(adjC) == board[r].charAt(c)
          && !connectedCells.contains(new Point(adjR, adjC))) {
        search(connectedCells, adjR, adjC);
      }
    }
  }
}

record Point(int r, int c) {}
